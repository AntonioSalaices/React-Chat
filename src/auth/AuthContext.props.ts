import { Func } from '@Utils/types';
import { User } from '@Models/Core';

export interface AuthError {
  statusCode: number;
  message: string;
}

export interface AuthContextProps {
  user: User | null;
  login: Func;
  logout: Func;
  loading: boolean;
  error: AuthError;
}
