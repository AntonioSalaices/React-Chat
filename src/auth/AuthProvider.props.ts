export interface AuthContextProviderProps {
  /**
   * Children node
   */
  children: JSX.Element | JSX.Element[];
}
