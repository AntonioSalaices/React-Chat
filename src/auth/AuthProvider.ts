/**
 * Date 30/06/23
 *
 * @author Antonio Salaices
 */

const token = 'cookieMethod';

const isAuthenticated = () => {
  return Boolean(token);
};

const auth = {
  isAuthenticated,
};

export { auth };
