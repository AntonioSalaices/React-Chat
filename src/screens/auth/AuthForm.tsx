import { HTMLType } from '@Constans/htmlConstants';
import { translate } from '@Translate/translate';
import { AuthFormProps } from './AuthForm.props';
import { Container, FormField } from '@Components/Core';
import { useRef, useState } from 'react';
import { FaRegEye, FaRegEyeSlash } from 'react-icons/fa';
import { usePostQuery } from '@Hooks/usePost/usePost';

const { VITE_BASE_API_URL, VITE_BASE_AUTH_URL } = import.meta.env;

const AuthForm: React.FC<AuthFormProps> = () => {
  const { responseData, loading, error, post } = usePostQuery(`${VITE_BASE_API_URL}${VITE_BASE_AUTH_URL}`);
  const [isShownPassword, setIsShownPassword] = useState<boolean>(false);
  const email = useRef<HTMLInputElement>(null);
  const password = useRef<HTMLInputElement>(null);
  console.log('response', responseData, loading, error);
  const onSubmit = (e) => {
    e.preventDefault();
    const body = {
      email: email?.current?.value,
      password: password?.current?.value,
    };
    post(body);
  };

  return (
    <Container>
      <div className="col justify-center align-center">
        <FormField
          ref={email}
          type={HTMLType.text}
          subTx="subPlaceholder.email"
          tx="form.inputs.email"
          testID="emailInput"
        />
        <FormField
          ref={password}
          type={isShownPassword ? HTMLType.text : HTMLType.password}
          tx="form.inputs.password"
          testID="passwordInput"
          right={
            isShownPassword ? (
              <FaRegEyeSlash className="pointer" onClick={() => setIsShownPassword(false)} />
            ) : (
              <FaRegEye className="pointer" onClick={() => setIsShownPassword(true)} />
            )
          }
        />
        <button className="btn-primary text-white" onClick={onSubmit}>
          {translate('buttons.signIn')}
        </button>
      </div>
    </Container>
  );
};

export default AuthForm;
