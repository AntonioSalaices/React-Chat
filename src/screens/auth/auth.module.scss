@import '../../assets/styles/palette';
@import '../../assets/styles/fonts';
@import '../../assets/styles/utilities';

.modalBackground {
  background-color: $background-color;
  width: 100vw;
  height: 100vh;
  z-index: 0;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  position: absolute;
}

.centered {
  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}

.modal {
  z-index: 10;
  width: 50vw;
  height: 50vh;
  background-color: $white;
  border-radius: $base-border-radius;
}
