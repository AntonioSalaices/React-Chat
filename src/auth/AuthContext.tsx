import { createContext, useEffect, useState } from 'react';
import { isEmpty } from 'lodash';

import { Func } from '@Utils/types';
import { Security } from '@Constans/securityConstants';
import { User, UserKeys } from 'models/User';
import { AuthContextProviderProps } from './AuthProvider.props';

import { usePostQuery } from '@Hooks/usePost/usePost';

const { VITE_BASE_API_URL, VITE_AUTH_LOGIN_URL, VITE_AUTH_LOGOUT_URL } = import.meta.env;

interface AuthContextProps {
  user: User | null;
  login: Func;
  logout: Func;
  loading: boolean;
}

export const AuthContext = createContext<AuthContextProps>({
  user: {} as User,
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  login: () => {},
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  logout: () => {},
  loading: false,
});

export const AuthProvider: React.FC<AuthContextProviderProps> = ({ children }) => {
  const [user, setUser] = useState<User | null>(() => {
    const user = localStorage.getItem(Security.UserData);

    if (user) return JSON.parse(user);

    return null;
  });

  const {
    responseData: responseLogin,
    loading: isLoadingLogin,
    post: loginPost,
  } = usePostQuery<User, User>(`${VITE_BASE_API_URL}${VITE_AUTH_LOGIN_URL}`);

  const { post: logoutPost } = usePostQuery(`${VITE_BASE_API_URL}${VITE_AUTH_LOGOUT_URL}`);

  useEffect(() => {
    const saveResponse = () => {
      if (!isEmpty(responseLogin) && UserKeys.Email in responseLogin && UserKeys.Id in responseLogin) {
        const userData = responseLogin as User;
        setUser(userData);
        localStorage.setItem(Security.UserData, JSON.stringify(userData));
      }
    };
    saveResponse();
  }, [responseLogin]);

  const login = (credentials: User) => {
    loginPost(credentials);
  };

  const logout = () => {
    logoutPost();
    setUser(null);
    localStorage.removeItem(Security.UserData);
  };

  return (
    <AuthContext.Provider value={{ user, login, logout, loading: isLoadingLogin }}>{children}</AuthContext.Provider>
  );
};
