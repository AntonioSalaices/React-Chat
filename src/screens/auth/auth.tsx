import AuthForm from './AuthForm';
import { Modal } from '@Components/Core';
import usePortal from '@Hooks/usePortal/usePortal';
import { Func } from '@Utils/types';

interface AuthModalProps {
  isShown: boolean;
  onClose: Func;
}

export const AuthModal: React.FC<AuthModalProps> = ({ isShown, onClose }) => {
  const Portal = usePortal(document.getElementById('portal-root') as HTMLElement);

  return (
    <>
      {isShown && (
        <Portal>
          <Modal onClose={onClose}>
            <AuthForm />
          </Modal>
        </Portal>
      )}
    </>
  );
};
AuthModal.displayName = 'AuthModal';
