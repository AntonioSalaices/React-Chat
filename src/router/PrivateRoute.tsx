import { Layout } from '@Components/Core';
import { AuthContext } from 'auth/AuthContext';
import { useContext } from 'react';
import { Navigate } from 'react-router-dom';

const PrivateRoute = () => {
  const { user } = useContext(AuthContext);

  return user ? <Layout /> : <Navigate to={'/landing-page'} replace />;
};

export default PrivateRoute;
