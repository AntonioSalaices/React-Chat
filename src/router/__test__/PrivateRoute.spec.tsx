import { RenderResult, render } from '@testing-library/react';
import { AuthProvider } from 'auth/AuthContext';
import { BrowserRouter } from 'react-router-dom';
import PrivateRoute from 'router/PrivateRoute';

describe('<PrivateRoute />', () => {
  let component: RenderResult;
  beforeEach(() => {
    component = render(
      <BrowserRouter>
        <AuthProvider>
          <PrivateRoute />
        </AuthProvider>
      </BrowserRouter>
    );
  });

  it('Should render outlet', () => {
    const { container } = component;

    expect(container).toMatchSnapshot();
  });
});
