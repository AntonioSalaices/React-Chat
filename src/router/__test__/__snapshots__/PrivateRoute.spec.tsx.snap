// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PrivateRoute /> Should render outlet 1`] = `<div />`;
